### ISSUE TITLE
[Numerical Bug] Loss may be NAN during training

### ISSUE THREAD

**Issue by Justobe** [chocoluffy/Tensorflow-in-MachineLearning#1]
Hi~

Thank you very much for sharing this code!

However, I found that loss may be NAN after some iterations (69572 iterations in my laptop). After carefully checking the code, I found that the following code may trigger NAN in loss:

In `Tensorflow-in-MachineLearning/softmax/softmax.py`:

    cross_entropy = -tf.reduce_sum(y_*tf.log(y))

If  y contains 0 (output of softmax ), the result of `tf.log(y)` is `inf` because `log(0)` is illegal , and this may cause the result of loss to become NAN.

It could be fixed by making the following changes:

    cross_entropy = -tf.reduce_sum(y_*tf.log(y + 1e-10))

or

    cross_entropy = -tf.reduce_sum(y_*tf.log(tf.clip_by_value(y,1e-10,1.0)))

Hope to hear from you. 

Thanks in advance!  : )

**Comment by Justobe** [comment]
@chocoluffy  :)
